ReactNativeClient/lib/models/Tag.js

```diff
@@ -36,18 +42,48 @@ class Tag extends BaseItem {
 		);
 	}
 
+	static async childrenTagIds(parentId) {
+		const rows = await this.db().selectAll('SELECT id FROM tags WHERE parent_id = ?', [parentId]);
+		return rows.map(r => r.id);
+	}
+
 	// Untag all the notes and delete tag
-	static async untagAll(tagId) {
+	static async untagAll(tagId, options = null) {
+		if (!options) options = {};
+		if (!('deleteChildren' in options)) options.deleteChildren = true;
+
+		const tag = await Tag.load(tagId);
+		if (!tag) return; // noop
+
+		if (options.deleteChildren) {
+			const childrenTagIds = await Tag.childrenTagIds(tagId);
+			for (let i = 0; i < childrenTagIds.length; i++) {
+				await Tag.untagAll(childrenTagIds[i]);
+			}
+		}
+
 		const noteTags = await NoteTag.modelSelectAll('SELECT id FROM note_tags WHERE tag_id = ?', [tagId]);
 		for (let i = 0; i < noteTags.length; i++) {
 			await NoteTag.delete(noteTags[i].id);
 		}
 
-		await Tag.delete(tagId);
+		// We're recursing over all children already, so this call doesn't have to be recursive.
+		await Tag.delete(tagId, { deleteChildren: false });
 	}
 
 	static async delete(id, options = null) {
 		if (!options) options = {};
+		if (!('deleteChildren' in options)) options.deleteChildren = true;
+
+		const tag = await Tag.load(id);
+		if (!tag) return; // noop
+
+		if (options.deleteChildren) {
+			const childrenTagIds = await Tag.childrenTagIds(id);
+			for (let i = 0; i < childrenTagIds.length; i++) {
+				await Tag.delete(childrenTagIds[i]);
+			}
+		}
```

Comment:
Are there cases where we don't want to delete the children? If not, I think we can also remove the option from here.

Reply:
Actually the purpose of `deleteChildren` in Folder.delete is to prevent it from happening in the synchronizer. Because in general synchronizer actions should not have side-effects or it can lead to very strange issue, synchronisation loops, etc. So if you add it there, you should also use that option where relevant in synchronizer.js

Reply:
That makes sense. I've created the `deleteChildren` option to match the interface of `Folder.delete` since they are alike. I missed its use in `synchronizer.js` - thanks, I'll look into it.